### ISSUE TITLE
Gradient of MvNormal 

### ISSUE THREAD

**Issue by paschermayr** [FluxML/Zygote.jl#353]
Hi guys, 

Working on Julia 1.2, Zygote 0.3.4.

I would like to evaluate several distributions in a summary function and get the gradients of it. Doing this in Flux.Tracker is fine, but I somehow do not manage to make it work in Zygote. Some distributions work, some do not. One example is MvNormal:

```
using Pkg;
pkg"add Zygote#master"
using Flux.Tracker

function f(x)
    logpdf(MvNormal([2.,3], 1.), x )
end
f([2.,2.])

Tracker.gradient(x -> f(x), [4.,5.]) #works
Zygote.gradient(x -> f(x), [4.,5.]) #not working - MethodError: no method matching transpose(::ScalMat{Float64})
```

**Comment by MikeInnes** [comment]
What's `ScalMat`? If it's a matrix type, it might be easy to fix this just by adding a transpose impl for it.

**Comment by jkbest2** [comment]
`ScalMat` is a matrix type (specifically a diagonal matrix with all elements the same magnitude) from the [`PDMats` package](https://github.com/JuliaStats/PDMats.jl), which is used by `Distributions` for all multivariate normal covariance/precision matrices. Is the preferred solution adding a `transpose` method for these matrix types in the `PDMats` package, or adding something on the `Zygote` side (I'm not sure what a "transpose impl" is)?

**Comment by MikeInnes** [comment]
Yeah, I just meant implementing `transpose` for ScalMat, which seems to fix the first issue at least.

It's interesting that this works for Tracker, and that should mean it's easy to get it working with Zygote. The error you posted is a bit tricky though: it comes from treating `ScalMat` as both a struct via `.` field access (gradient is a NamedTuple with a gradient for each field) and as an array (gradient is an array). To fix this you'd have to figure out how the `ScalMat` is being used exactly.

Registering Distributions.jl's `gradlogpdf` implementations is definitely another option.

**Comment by xukai92** [comment]
How can I register Distributions.jl's `gradlogpdf`?

**Comment by MikeInnes** [comment]
We'd have to either add a dependency on Distributions here or add ZygoteRules to Distributions, and then define the adjoint for `logpdf`.

**Comment by fedario** [comment]
Any news on this or some way I can help?
Having this would be really convenient for variational inference.

**Comment by mohamed82008** [comment]
Check out https://github.com/TuringLang/DistributionsAD.jl.

**Comment by fedario** [comment]
Thanks, I'll try it!
Fun fact: I just saw you referring to that package here https://github.com/FluxML/Zygote.jl/issues/521 after trying univariate normals and getting the next error. :smile:

**Comment by CarloLucibello** [comment]
MvNormal gradients are provided by DistributionsAD.jl now. Closing as resolved-upstream.

_(Triaged against Zygote v0.7.10 on Julia 1.12.6.)_
